src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,394 @@
+
+Property Success As %Status [ InitialExpression = 0, Required ];
+
+Property Committed As %Boolean [ InitialExpression = 0, Required ];
+
+Property CommandString As %String(MAXLEN = 8192) [ Required ];
+
+ClassMethod Init(Action As %String, Package As %IPM.DataType.ModuleName) As %IPM.General.History [ Private ]
+{
+    Set log = ..%New()
+    Set log.TimeStart = $ZDateTime($Now(), 3)
+    Set log.Action = Action
+    Set log.Package = Package
+    Set log.CommandString = $Get($$$ZPMCommandToLog, "<unset>")
+    Set log.NameSpace = $Namespace
+    Set log.UserName = $Username
+    $$$ThrowOnError(log.SetVersion()) // Set the version to placeholder 0.0.0-0, Will be updated before finalizing
+    $$$ThrowOnError(log.%Save()) // Save it now in case Finalize isn't called
+    Quit log
+}
+
+ClassMethod InstallInit(Package As %IPM.DataType.ModuleName) As %IPM.General.History
+{
+    Quit ..Init("install", Package)
+}
+
+ClassMethod LoadInit() As %IPM.General.History
+{
+    // Package name is not known at this point, so use a placeholder
+    Quit ..Init("load", "")
+}
+
+ClassMethod UninstallInit(Package As %IPM.DataType.ModuleName) As %IPM.General.History
+{
+    Quit ..Init("uninstall", Package)
+}
+
+Method SetSource(ServerName As %String) As %Status
+{
+    $$$ThrowOnError(..%Reload())
+    Set server = ##class(%IPM.Repo.Definition).ServerDefinitionKeyOpen(ServerName)
+    If '$IsObject(server) {
+        Quit $$$ERROR($$$GeneralError,$$$FormatText("Repository '%1' is not defined.", ServerName))
+    }
+    Set ..SourceName = ServerName
+    Set ..SourceMoniker = server.#MONIKER
+    Set ..SourceDetails = server.Details
+    Quit ..%Save()
+}
+
+Method SetVersion(pVersion As %IPM.General.SemanticVersion) As %Status
+{
+    If ..%Id() '= "" {
+        $$$ThrowOnError(..%Reload())
+    }
+    If $Get(pVersion) = "" {
+        Set pVersion = "0.0.0-0"
+    }
+    Set ..Version = $Select($IsObject(pVersion): pVersion, 1: ##class(%IPM.General.SemanticVersion).FromString(pVersion))
+    Quit ..%Save()
+}
+
+Method SetName(pName As %String) As %Status
+{
+    $$$ThrowOnError(..%Reload())
+    Set ..Package = pName
+    Quit ..%Save()
+}
+
+Method Finalize(status As %Status, dev As %Boolean = 0) As %Status
+{
+    $$$ThrowOnError(..%Reload())
+    Set ..Success = status
+    Set ..Committed = (dev = 1) || $$$ISOK(status)
+    Set ..TimeEnd = $ZDateTime($Now(), 3)
+    If ..Package = "" {
+        Return $$$ERROR($$$GeneralError, "Package name not set")
+    }
+    Return ..%Save()
+}
+
+/// @API.Method
+/// Get the history of all installations, uninstalls, and loads in given namespace
+/// The filter argument is a multidimensional array with structure 
+/// filter(columnName) = value
+/// Where value can optionally start with >, >=, <, <=, =, <> or contain * to indicate a wildcard
+ClassMethod GetHistory(ByRef filter, ascend As %Boolean = 0, limit As %Integer = 0, namespace As %String) As %SQL.StatementResult
```

Comment:
If we're going to make an API method that returns a %SQL.StatementResult, I'd rather expose it as a class query.